### ISSUE TITLE
behavior of `localize_vars` and variables within top-level `if` statements

### ISSUE THREAD

**Issue by amitmurthy** [JuliaLang/julia#12367]
With `julia -p4`

```
if true
    n = 10^4
    for (idx,p) in enumerate(procs())
        @spawnat p begin
            @sync for i in 1:n
                ;
            end
        end
    end
end
```

throws exceptions on the workers. 

```
julia> exception on 2: ERROR: UndefVarError: n not defined
 in anonymous at multi.jl:347
 in anonymous at multi.jl:877
 in run_work_thunk at multi.jl:619
 in run_work_thunk at multi.jl:628
 in anonymous at task.jl:11
```

Without the `if` condition, i.e. just

```
    n = 10^4
    for (idx,p) in enumerate(procs())
        @spawnat p begin
            @sync for i in 1:n
                ;
            end
        end
    end
```

it works fine.

~~I am fairly certain that this was not a problem before.~~ Apparently it always was.

**Comment by tkelman** [comment]
This fails for me even on 0.3.10. Can you find a commit where it worked?

**Comment by amitmurthy** [comment]
I thought it was working a couple of weeks back, tried a random commit from that period, but no, it does not.

Next question then - it should work, right? Will leave it open as a bug.

**Comment by amitmurthy** [comment]
Closed by #19594

**Comment by tkelman** [comment]
is this tested?
